pkg/hs/urbit-king/lib/Urbit/Vere/Ports.hs

```diff
@@ -30,12 +31,24 @@ buildInactivePorts = PortControlApi noop noop
  where
   noop x = pure ()
 
+data TryNat
+  = TryNatAlways
+  | TryNatWhenPrivate
+
 -- | Builds a PortControlApi struct which tries to hole-punch by talking to the
 -- NAT gateway over NAT-PMP.
 buildNatPorts :: (HasLogFunc e)
-              => (Text -> RIO e ())
+              => TryNat
+              -> (Text -> RIO e ())
               -> RIO e PortControlApi
-buildNatPorts stderr = do
+
+buildNatPorts TryNatWhenPrivate stderr = do
+  behind <- likelyBehindRouter
+  if behind
+    then buildNatPorts TryNatAlways stderr
```

Comment:
The recursion feels a bit strange here. Like, we're recursing on the structure of TryNat, but it doesn't actually have recursive structure. This is potentially another argument in favor of having three functions: buildInactivePorts, buildNatPorts, buildNatPortsWhenPrivate. The fact that the last one conditionally calls the other two now feels appropriate, since it's delegation, rather than recursion.

This is not a huge issue, but I'm trying to give you a sense of the haskell taste for these things.

Reply:
Went with 3 functions.